File: python/visp/detector.py
import cv2
import numpy

import _visp


class QRCodeDetector(_visp.QRCodeDetector):
    def detect(self, img):
        return [QRCode(c) for c in _visp.QRCodeDetector.detect(self, img)]


class QRCode(object):
    def __init__(self, code):
        self._code = code

    def draw(self, img):
        cv2.polylines(img, numpy.array([self.polygon], dtype=numpy.int32),
                      True, (0, 0, 255), 2)

        cv2.putText(img, self.message, self.center,
                    cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)

    @property
    def message(self):
        return self._code.message

    @property
    def center(self):
        return tuple(map(int, self._code.cog))

    @property
    def bounding_box(self):
        return numpy.array(self._code.bounding_box)

    @property
    def polygon(self):
        return numpy.array(self._code.polygon)

File: _visp/detector.cpp
#include <png.h>

#include "detector.h"

#include <visp/vpImageConvert.h>

#include "VISPBoostConverter.h"


VispQRCodeDetector::VispQRCodeDetector() {
}

list VispQRCodeDetector::detect(const cv::Mat &img) {
  vpImage<unsigned char> I;
  vpImageConvert::convert(img, I);

  list qrCodes;

  if (detector.detect(I)) {
    for (uint i=0; i < detector.getNbObjects(); i++) {

      QRCode code;
      code.polygon = VISPConverter::convertToBoost(detector.getPolygon(i));
      code.message = detector.getMessage(i);
      code.bounding_box = VISPConverter::convertToBoost(detector.getBBox(i));
      code.cog = VISPConverter::convertToBoost(detector.getCog(i));

      qrCodes.append(code);
    }
  }

  return qrCodes;
}
